urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/geom/twodim/ops/IntersectSimplePolygon.java

```diff
@@ -0,0 +1,231 @@
+package coho.geom.twodim.ops;
+
+import java.util.*;
+import coho.common.number.*;
+import coho.geom.twodim.*;
+
+class IntersectSimplePolygon {
+	private ArrayList<IntersectSegment[]> hullEdges = new ArrayList<IntersectSegment[]>();
+	private IntersectPoint left;//, right;//the left and right most point
+	private BoundingBox bbox;//bounding box for this polgyon
+	public IntersectPoint left(){
+		return left;
+	}
+//	public IntersectPoint right(){
+//		return right;
+//	}
+	public BoundingBox bbox(){
+		return bbox;
+	}
+	public int hulls(){
+		return hullEdges.size();
+	}
+	public IntersectSegment[] hull(int i){
+		return hullEdges.get(i);
+	}
+	public IntersectSimplePolygon(ConvexPolygon p){
+		this.bbox = p.bbox();
+		int n = p.degree();
+		
+		//compute upper/lower hull;
+		int ll=0, ur=0; //left lower, right upper point 
+		for(int i=1; i<n; i++){
+			if(p.point(i).compareTo(p.point(ll))<0){
+				ll = i;
+			}
+			if(p.point(i).compareTo(p.point(ur))>0){
+				ur = i;
+			}
+		}
+		//break into lower and upper half. 
+		//Because polygon is anti-clock-wise, lower is ll->ur
+        //FIXED: don't use fwd and bwd. Assume polygon is anti-clock-wise. 
+		//If support both, use direction of edges to tell which is upper/lower
+		int lowerLength = (ur-ll)+1;  
+		int upperLength = (n+ll-ur)+1;
+		if(ur<ll){
+			lowerLength = (n+ur-ll)+1;
+			upperLength = (ll-ur)+1;
+		}
+		Point[] lowerPoints = new Point[lowerLength];
+		Point[] upperPoints = new Point[upperLength];
+		for(int i=0; i<lowerLength; i++){
+			lowerPoints[i] = p.point((ll+i)%n);//fwd
+		}
+		for(int i=0; i<upperLength; i++){
+			upperPoints[i] = p.point((ll-i+n)%n);//bwd
+		}
+		
+		//Create IntersectSegment for each edge. 
+		IntersectSegment[] lowerSegs = new IntersectSegment[lowerPoints.length-1];
+		for(int i=0; i<lowerSegs.length;i++){
+			lowerSegs[i] = new IntersectSegment(Segment.create(lowerPoints[i],lowerPoints[i+1]));//anti-clock wise order
+		}
+		IntersectSegment[] upperSegs = new IntersectSegment[upperPoints.length-1];
+		for(int i=0; i<upperSegs.length; i++){
+			upperSegs[i] = new IntersectSegment(Segment.create(upperPoints[i+1], upperPoints[i]));//anti-clock wise order
+		}
+		
+		//Create IntersectPoint for each vertex. Store it in IntersectSegment
+		IntersectSegment pre = upperSegs[0];
+		IntersectSegment pos = lowerSegs[0];
+		left = new IntersectPoint(lowerPoints[0],pre,pos,true);//exact value
+		IntersectPoint pi = left;
+		pre.add(pi);
+		pos.add(pi);
+		
+		for(int i=1; i<lowerSegs.length;i++){//lower hull
+			pre = lowerSegs[i-1];
+			pos = lowerSegs[i];
+			pi = new IntersectPoint(lowerPoints[i], pre,pos, true);
+			pre.add(pi);
+			pos.add(pi);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 77:83, 92:98.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.